A bunch of stuff for profile and resources

macdude95 opened the pull request:
Stuff is kinda ugly, especially my custom thing... but everything is functional i believe
<img width="624" alt="screen shot 2018-07-05 at 7 37 12 pm" src="https://user-images.githubusercontent.com/22382440/42357097-00b06d6e-808b-11e8-97c7-54dff5a681dc.png">
<img width="624" alt="screen shot 2018-07-05 at 7 37 05 pm" src="https://user-images.githubusercontent.com/22382440/42357098-00c9048c-808b-11e8-8797-88a19b0d8354.png">
<img width="624" alt="screen shot 2018-07-05 at 7 36 54 pm" src="https://user-images.githubusercontent.com/22382440/42357099-00e0f2c2-808b-11e8-982c-e400fcf5e11c.png">
<img width="624" alt="screen shot 2018-07-05 at 7 36 48 pm" src="https://user-images.githubusercontent.com/22382440/42357100-00f79d6a-808b-11e8-93e7-0db76e75bf39.png">
<img width="624" alt="screen shot 2018-07-05 at 7 36 43 pm" src="https://user-images.githubusercontent.com/22382440/42357101-01218224-808b-11e8-8318-1913c7e42dc5.png">

Commit:
fixed ministry team details storyboard again

Commit:
made the selector in the crusegmented control move more smothely and have interpolating colors

Commit:
Profile view hides when you switch tabs

Commit:
now adds tab bar changed obserbver for popping the profile view automatically to vcs with profile button

Commit:
started creating resources firebase stuff

Commit:
now resources data is pulled from database

Commit:
added detail view controllers for resources

Commit:
started working on mini audio player for resources screen

Commit:
some more fine tuning of resources tab

Commit:
formatted resource dates and now doesn't show image if no image

Commit:
added a resource manager so i don't redownload resources all the time

Commit:
added refresh functionality

Commit:
fewer hard coded values and now we have play and pause button

Commit:
changed play and pause button color

Commit:
added constraints to an activity indicator

tydahlwave commented on CruCentralCoast/Info.plist:
```diff
@@ -2,6 +2,11 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
+	<key>NSAppTransportSecurity</key>
```
What was this needed for?

macdude95 replied:
Pretty sure that was needed to allow my web views in the resources section to load any website, even if it is not https.

tydahlwave replied:
👍

tydahlwave commented on CruCentralCoast/CustomViews/CruAudioControl.swift:
```diff
@@ -0,0 +1,101 @@
+}
+
+extension CruAudioControl : AVPlayerViewControllerDelegate {
+    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
```
How is this being used? I don't see where it will be called

macdude95 replied:
I dont call this method explicitly. I add an observer to the AVPlayerViewController i create, and it is called automatically when the value at 
```swift
#keyPath(AVPlayerViewController.view.frame)
``` 
changes. I do this because i couldn't think of any other way to detect if that view controller is dismissed.

tydahlwave replied:
I'm still a little confused what you mean. Why do you need to know when a view controller is dismissed?

macdude95 replied:
I wanted to know so that i can the audio pause automatically when you close it... but maybe that's not necessary... i might just get rid of that. it would make the code more simple and it would also maybe make the user experience more simple.

tydahlwave commented on CruCentralCoast/CustomViews/CustomSegmentedControl.swift:
```diff
@@ -111,20 +112,38 @@ class CruSegmentedControl: UIControl {
+        for button in self.buttons {
+            // is the selector between me and another button? (within a certain ammount)
+            let centerOfButton = button.frame.origin.x + button.frame.width/2
+            if centerOfButton - selectorWidth ... centerOfButton + selectorWidth ~= centerOfSelector {
```
Wow, nice use of advanced operators. I had to look up ... and ~=

macdude95 replied:
Haha tbh i had to go look them up again too. I wrote this code a WHILE ago. However, I can't take credit for coming up with the idea to use this myself. I was following a tutorial =P

tydahlwave replied:
👌

tydahlwave commented on CruCentralCoast/Database/Resource.swift:
```diff
@@ -0,0 +1,75 @@
+    }
+    
+    required init?(dict: NSDictionary) {
+        guard let author = dict["author"] as? String else {
```
Would it be cleaner to perform all of the guard lets prior to setting any values? Because technically if a value is set and a later guard let fails, then returning nil negates the fact that the value was set

macdude95 replied:
Yes you are correct. Ill fix that.

tydahlwave commented on CruCentralCoast/MainTabBarController.swift:
```diff
@@ -0,0 +1,22 @@
+    static let tabBarChangedNotification = Notification.Name(MainTabBarController.tabBarChangedString)
+    
+    override func tabBar(_ tabBar: UITabBar, didSelect item: UITabBarItem) {
+        // this works since self.selectedIndex isn't updated until after this call
```
Nice comment

tydahlwave commented on CruCentralCoast/Extensions/UIViewControllerExtensions.swift:
```diff
@@ -48,4 +54,25 @@ public extension UIViewController {
+            let activityIndicator = UIActivityIndicatorView(activityIndicatorStyle: .whiteLarge)
+            activityIndicator.hidesWhenStopped = true
+            activityIndicator.color = .gray
+            webView.addSubview(activityIndicator)
```
Does this activity indicator stop animating? I don't see that code

tydahlwave replied:
Now I see that it is the navigation delegate's job to stop animating it, which makes sense

tydahlwave replied:
Although, I wonder if it would be possible to assign the UIActivityIndicatorView to a variable in the navigation delegate so that later you don't have to find it by searching through the subviews?

macdude95 replied:
good idea. ill change some stuff

tydahlwave commented on CruCentralCoast/Extensions/UIViewControllerExtensions.swift:
```diff
@@ -6,13 +6,15 @@
         if let largeTitleView = self.navigationController?.navigationBar.subviews.first(where: {
             String(describing: type(of: $0)) == "_UINavigationBarLargeTitleView" } ) {
             
+            NotificationCenter.default.addObserver(self, selector: #selector(self.popToRootViewController), name: MainTabBarController.tabBarChangedNotification, object: nil)
```
Is this method the right place for this?

macdude95 replied:
Would you maybe recommend doing it in the AppDelegate?

tydahlwave replied:
I feel like the best place would be in the viewDidLoad() method on each tab's root view controller

macdude95 replied:
Oh yeah duh that's smart. ill do that

macdude95 replied:
i think the original reason i did it here was to make it so i didn't need to put that line of code in each individual view controller.  however, i agree that it doesn't really seem to fit in this method. I went ahead and moved it.

tydahlwave commented on CruCentralCoast/Database/DatabaseManager.swift:
```diff
@@ -33,6 +33,38 @@ class DatabaseManager {
         getData(completion)
     }
+    
+    private func getArticleResources(_ completion: @escaping ([ArticleResource])->Void) {
```
Is this method ever used? Same with next two methods

macdude95 replied:
I think that was leftover stuff from something else i used to do =P ill get rid of those. good catch

tydahlwave commented on CruCentralCoast/Resources/Articles/ArticleResourceDetailVC.swift:
```diff
@@ -0,0 +1,41 @@
+
+extension ArticleResourceDetailVC: WKNavigationDelegate {
+    func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!) {
+        if let activityIndicator = webView.subviews.first(where: { String(describing: type(of: $0)) == "UIActivityIndicatorView" }) as? UIActivityIndicatorView {
```
Can the activity indicator be added as a variable on the WKNavigationDelegate so we don't have to search through the subviews for it?

macdude95 replied:
i mean there is some clever way to do stored properties in extensions, but i'm not trying to deal with that magic. Instead, I just stored it in the viewcontroller class itself. about to push some code.

tydahlwave commented on CruCentralCoast/Resources/Audio/AudioResourceDetailVC.swift:
```diff
@@ -0,0 +1,33 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.navigationItem.title = resource?.title
```
self.

macdude95 replied:
oops

tydahlwave commented on CruCentralCoast/Resources/ResourcesTableViewCollectionViewCell.swift:
```diff
@@ -30,56 +34,64 @@ class ResourcesTableViewCollectionViewCell: UICollectionViewCell {
+            cell.dateLabel.text = resource.formattedDate
+            if let videoResource = resource as? VideoResource {
+                if let url = URL(string: videoResource.imageURL) {
+                    cell.previewImage.downloadedFrom(url: url)
```
Will this show the incorrect image if the cell is reused while loading the image?

macdude95 replied:
I'm actually not sure... that's a great question... possibly yet another reason to have a completely separate image manager that deals with all of this image downloading stuff rather than having the cell or imageview itself handling the downloading.... maybe...

tydahlwave replied:
True, I guess we can fix it later when we create an image manager

tydahlwave commented on CruCentralCoast/Resources/ResourcesTableViewCollectionViewCell.swift:
```diff
@@ -30,56 +34,64 @@ class ResourcesTableViewCollectionViewCell: UICollectionViewCell {
+    
+    @objc private func refreshResources(_ refreshControl: UIRefreshControl) {
+        ResourceManager.instance.refreshResources(ofType: self.type) { resources in
+            UIView.animate(withDuration: 0.3) {
```
Why is this animated? What gets animated?

macdude95 replied:
!!! you just fixed this stupid UI bug i was struggling with for way too long. I have no clue why i put this inside a 
```swift
UIView.animate(withDuration:)
```
block. But because i did, i was getting this weird effect where the activity indicator was overlapping the tableview content. But getting rid of the call to ```UIView.animate()``` just fixed it! thank you!

tydahlwave replied:
🙌

tydahlwave commented on CruCentralCoast/Resources/ResourceManager.swift:
```diff
@@ -0,0 +1,43 @@
+        if retrievedResources.contains(type) {
+            completion(self.resources.filter { $0.type == type })
+        }
+        else {
```
else { should be on same line as }

macdude95 replied:
oops.

tydahlwave commented on CruCentralCoast/Resources/ResourcesVC.swift:
```diff
@@ -89,3 +97,69 @@ extension ResourcesVC: UICollectionViewDelegateFlowLayout {
+    func showAVPlayerViewController(player: AVPlayer) {
+        let vc = AVPlayerViewController()
+        vc.player = player
+        vc.addObserver(self.audioController, forKeyPath: #keyPath(AVPlayerViewController.view.frame), options: [], context: nil)
```
What is this observer used for?

macdude95 replied:
I think i addressed this in one of my replies to one of your other comments. but if you're still confused let me know.

tydahlwave reviewed (changes requested):
Very nice code! So many beautiful design patterns. I posted some comments that I was curious about. Also, I think the constraints on the audio popup still need a little work, but very close (Shouldn't it be closer to the bottom of the screen? I think the time labels are not aligned).

Commit:
fixed stuff according to Tyler's helpful comments on my pull request

Commit:
Merge branch 'release_2.0' into michael

macdude95 commented:
Pushed some changes, but there are a couple more things i want to deal with before this branch is merged in.

Commit:
removed some unnecesary files (red files in heirarchy)

Commit:
added in ministryteamdetails.storyboard

macdude95 commented:
<img width="520" alt="screen shot 2018-07-08 at 1 10 41 pm" src="https://user-images.githubusercontent.com/22382440/42423564-245aeeba-82b1-11e8-9725-0788130ffa26.png">

fixed the constraints on the mini audio player

macdude95 commented:
And with that, I think i addressed all of your comments (besides the image downloading one, but i dont think that's gonna be that big of a deal anyway right now because i dont know how often the database will even have thumbnail images for those cells so that case will rarely be entered.)

Commit:
fixed constraint of mini audio player

Commit:
i think i finished the changes for this pull request

Commit:
removed extra constraint on audio player

tydahlwave committed:
Merge branch 'release_2.0' into michael
